### PR TITLE
[IR] Enforce type check for all expressions

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #3301

This PR adds `type_check` for remaining expressions and removes all `TODO: assert no unknowns after type_check for all expressions are implemented`.

### CODE REVIEW

File: taichi/ir/frontend_ir.cpp [hunk 1]
```diff
@@ -128,10 +131,12 @@ void UnaryOpExpression::serialize(std::ostream &ss) {
 }
 
 void UnaryOpExpression::type_check() {
-  // TODO: assert no unknowns after type_check for all expressions are
-  // implemented
-  if (operand->ret_type == PrimitiveType::unknown)
-    return;
+  TI_ASSERT_INFO(operand->ret_type != PrimitiveType::unknown,
```
Comment (ailzhang) [inline taichi/ir/frontend_ir.cpp:134]: nit: maybe this assert worth its own macro :P

**Review by ailzhang — APPROVED** [review]
LGTM!

File: taichi/ir/frontend_ir.cpp (hunk 1 above)
Comment (strongoier) [reply taichi/ir/frontend_ir.cpp]: Nice suggestion! Done.

**Pull request merged**